packages/roosterjs-editor-plugins/lib/plugins/TableResize/editors/TableEditor.ts

```diff
@@ -0,0 +1,207 @@
+import createTableResizer from './TableResizer';
+import TableEditFeature, { disposeTableEditFeature } from './TableEditorFeature';
+import { ChangeSource, IEditor } from 'roosterjs-editor-types';
+import { getComputedStyle, normalizeRect } from 'roosterjs-editor-dom';
+
+const INSERTER_HOVER_OFFSET = 5;
+
+/**
+ * @internal
+ *
+ * A table has 5 hot areas to be resized/edited (take LTR example):
+ *   [                ]
+ *  +[      1         ]+--------------------+
+ *  |[                ]|                    |
+ * [ ]               [ ]                    |
+ * [ ]               [ ]                    |
+ * [2]               [3]                    |
+ * [ ]               [ ]                    |
+ * [ ][       4       ]|                    |
+ *  +------------------+--------------------+
+ *  |                  |                    |
+ *  |                  |                    |
+ *  |                  |                    |
+ *  +------------------+--------------------+
+ *                                           [5]
+ *
+ * 1 - Hover area to show insert column button
+ * 2 - Hover area to show insert row button
+ * 3 - Hover area to show vertical resizing bar
+ * 4 - Hover area to show horizontal resizing bar
+ * 5 - Hover area to show whole table resize button
+ *
+ * When set a different current table or change current TD, we need to update these areas
+ */
+export default class TableEditor {
+    // 1, 2 - Insert a column or a row
+    private horizontalInserter: TableEditFeature;
+    private verticalInserter: TableEditFeature;
+
+    // 3, 4 - Resize a column or a row from a cell
+    private horizontalResizer: TableEditFeature;
+    private verticalResizer: TableEditFeature;
+
+    // 5 - Resize whole table
+    private tableResizer: TableEditFeature;
+
+    private isRTL: boolean;
+    private isChanged: boolean;
+
+    constructor(private editor: IEditor, public readonly table: HTMLTableElement) {
+        const sizeTransformer = editor.getSizeTransformer();
+        this.isRTL = getComputedStyle(table, 'direction') == 'rtl';
+        this.tableResizer = createTableResizer(
+            table,
+            sizeTransformer,
+            this.isRTL,
+            this.onEditTable
+        );
+        this.editor.addUndoSnapshot();
+    }
+
+    dispose() {
+        if (this.isChanged && !this.editor.isDisposed()) {
+            this.editor.addUndoSnapshot(null /*callback*/, ChangeSource.Format);
+        }
+
+        this.disposeTableResizer();
+        this.disposeCellResizers();
+        this.disposeTableInserter();
+    }
+
+    isTableChanged() {
+        return this.isChanged;
+    }
+
+    onMouseMove(x: number, y: number) {
+        for (let i = 0; i < this.table.rows.length; i++) {
+            const tr = this.table.rows[i];
+            let j = 0;
+            for (; j < tr.cells.length; j++) {
+                const td = tr.cells[j];
+                const tdRect = normalizeRect(td.getBoundingClientRect());
```

Comment:
I think we need a null conditional here, if table is too long there is an exception

Reply:
sounds good